src/implied_volatility.rs

```diff
@@ -8,117 +12,135 @@ where
     fn calc_rational_iv(&self) -> Result<f64, String>;
 }
 
-impl ImpliedVolatility<f32> for Inputs {
-    /// Calculates the implied volatility of the option.
-    /// Tolerance is the max error allowed for the implied volatility,
-    /// the lower the tolerance the more iterations will be required.
-    /// Recommended to be a value between 0.001 - 0.0001 for highest efficiency/accuracy.
-    /// Initializes estimation of sigma using Brenn and Subrahmanyam (1998) method of calculating initial iv estimation.
-    /// Uses Newton Raphson algorithm to calculate implied volatility.
-    /// # Requires
-    /// s, k, r, q, t, p
-    /// # Returns
-    /// f32 of the implied volatility of the option.
-    /// # Example:
-    /// ```
-    /// use blackscholes::{Inputs, OptionType, ImpliedVolatility};
-    /// let inputs = Inputs::new(OptionType::Call, 100.0, 100.0, Some(0.5), 0.05, 0.2, 20.0/365.25, None);
-    /// let iv = inputs.calc_iv(0.0001).unwrap();
-    /// ```
-    /// Initial estimation of sigma using Modified Corrado-Miller from ["A MODIFIED CORRADO-MILLER IMPLIED VOLATILITY ESTIMATOR" (2007) by Piotr P√luciennik](https://sin.put.poznan.pl/files/download/37938) method of calculating initial iv estimation.
-    /// A more accurate method is the "Let's be rational" method from ["Let’s be rational" (2016) by Peter Jackel](http://www.jaeckel.org/LetsBeRational.pdf)
-    /// however this method is much more complicated, it is available as calc_rational_iv().
-    #[allow(non_snake_case)]
-    fn calc_iv(&self, tolerance: f32) -> Result<f32, String> {
-        let mut inputs: Inputs = self.clone();
-
-        let p = self
-            .p
-            .ok_or("inputs.p must contain Some(f32), found None".to_string())?;
-        // Initialize estimation of sigma using Brenn and Subrahmanyam (1998) method of calculating initial iv estimation.
-        // commented out to replace with modified corrado-miller method.
-        // let mut sigma: f32 = (PI2 / inputs.t).sqrt() * (p / inputs.s);
-
-        let X: f32 = inputs.k * E.powf(-inputs.r * inputs.t);
-        let fminusX: f32 = inputs.s - X;
-        let fplusX: f32 = inputs.s + X;
-        let oneoversqrtT: f32 = 1.0 / inputs.t.sqrt();
-
-        let x: f32 = oneoversqrtT * (SQRT_2PI / (fplusX));
-        let y: f32 = p - (inputs.s - inputs.k) / 2.0
-            + ((p - fminusX / 2.0).powf(2.0) - fminusX.powf(2.0) / PI).sqrt();
-
-        let mut sigma: f32 = oneoversqrtT
-            * (SQRT_2PI / fplusX)
-            * (p - fminusX / 2.0 + ((p - fminusX / 2.0).powf(2.0) - fminusX.powf(2.0) / PI).sqrt())
-            + A
-            + B / x
-            + C * y
-            + D / x.powf(2.0)
-            + _E * y.powf(2.0)
-            + F * y / x;
-
-        if sigma.is_nan() {
-            Err("Failed to converge".to_string())?
-        }
+macro_rules! impl_iv {
```

Comment:
could you explain a bit why a macro is the best choice here?

Reply:
I believe the main reason to favor macros in this case is because using `<T as num_traits::NumCast>::from` to convert primitives or constants to a bounded Float type results in extra runtime instructions compared to explicitly implementing these functions for each float type, where the compiler will determine how to represent those values as literals for each type.

That said, I don't know for certain how smart the compiler is in this regard. I'm also sure the overhead is pretty small, but given that `lets_be_rational` is pretty well optimized I wanted to be conscious of that.

Secondarily, having to `T::from(foo).unwrap()` every literal is pretty ugly. Not sure if there's a nice way around that

Reply:
I take back the performance point, I just took a look at compiler explorer and it seems like it should be no extra overhead. So I can convert everything to trait bounds, but the secondary aesthetic point is still something that bothers me a little XD